tools/wake/timeline.cpp

```diff
@@ -0,0 +1,242 @@
+    JobNode &job = job_map.at(job_id);
+    job_map.at(job_id).top_sort_visited = true;
+    for (auto it = job.dependencies.begin(); it < job.dependencies.end(); it++) {
+        if (job_map.find(*it) == job_map.end()) {
+            job.dependencies.erase(it--);
+            continue;
+        }
+        if (!(job_map.at(*it).top_sort_visited)) {
+            dfs_top_sort(*it, job_map, top_sorted_jobs);
+        }
+    }
+    top_sorted_jobs.emplace_back(job_id);
+}
+
+void top_sort(std::map<long, JobNode> &job_map, std::vector<long> &top_sorted_jobs) {
+    for (const auto &job_pair: job_map) {
+        if (!(job_pair.second.top_sort_visited)) {
+            dfs_top_sort(job_pair.first, job_map, top_sorted_jobs);
+        }
+    }
+}
+
+long fill_one_dependency(std::pair<int, long> access, long dependency, std::map<long, JobNode> &job_map) {
+    if (access.first == 2) {
+        return access.second;
+    }
+    if (job_map.find(access.second) != job_map.end()) {
+        job_map.at(access.second).dependencies.emplace_back(dependency);
+    }
+    return dependency;
+}
+
+void fill_all_dependencies(const std::vector<std::pair<int, long>> &accesses, std::map<long, JobNode> &job_map) {
+    long dependency = -1;
+    for (auto access: accesses) {
+        dependency = fill_one_dependency(access, dependency, job_map);
+    }
+}
+
+void assign_parents(std::map<long, JobNode> &job_map, std::vector<long> &top_sorted_jobs) {
+    for (long job_id: top_sorted_jobs) {
+        JobNode &job_node = job_map.at(job_id);
+        for (long dependency_id: job_node.dependencies) {
+            JobNode dependency = job_map.at(dependency_id);
+            if (dependency.hypothetical_endtime >= job_node.hypothetical_starttime) {
+                job_node.hypothetical_starttime = dependency.hypothetical_endtime;
+                job_node.parent_id = dependency_id;
+            }
+        }
+
+        job_node.hypothetical_endtime = job_node.hypothetical_starttime + (job_node.endtime - job_node.starttime);
+    }
+}
+
+void find_critical_path(std::map<long, JobNode> &job_map, std::vector<long> &critical_path) {
+    long latest_job_id = -1;
+    for (const auto &job_pair: job_map) {
+        if (latest_job_id == -1 ||
+            job_map.at(latest_job_id).hypothetical_endtime < job_pair.second.hypothetical_endtime) {
+            latest_job_id = job_pair.second.job.job;
+        }
+    }
+    job_map.at(latest_job_id).is_on_critical_path = true;
+    critical_path.emplace_back(job_map.at(latest_job_id).job.job);
+    while (job_map.at(latest_job_id).parent_id != -1) {
+        latest_job_id = job_map.at(latest_job_id).parent_id;
+        job_map.at(latest_job_id).is_on_critical_path = true;
+        critical_path.emplace_back(job_map.at(latest_job_id).job.job);
+    }
+}
+
+void dfs_transitive_reduction(long job_id, std::map<long, JobNode> &job_map, std::vector<int> &ancestors) {
```

Comment:
Neat! Some how I made it through a CS education without learning about this! Gotta go do some reading.